Text index: introduce cache for the deserialized header

ahmadov opened the pull request:
<!--- Disable AI PR formatting assistant: true -->
### Changelog category (leave one):

- Improvement

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):

Introduced cache for the deserialized header of the text index to reduce I/O and improve query performance.
The cache can be configured via new server settings:
- `text_index_header_cache_policy`
- `text_index_header_cache_size`
- `text_index_header_cache_max_entries`
- `text_index_header_cache_size_ratio`

and enabled using `use_text_index_header_cache` setting (disabled by default).

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

### Details

The header of the text index contains the number of tokens, bloom filter, and sparse index. Since both bloom filter and sparse index can be relatively large, caching them reduces I/O and improves query performance.

This PR makes the cache configurable with the following new server settings:
- `text_index_header_cache_policy`
- `text_index_header_cache_size`
- `text_index_header_cache_max_entries`
- `text_index_header_cache_size_ratio`

Additionally, the header cache can be enabled via the new setting `use_text_index_header_cache`. By default, the cache is disabled.

ahmadov committed:
Text index: introduce cache for the deserialized header

The header of the text index contains the number of tokens, bloom filter
and sparse index.
Since both bloom filter and sparse index can be relatively large,
caching them would reduce I/O and improve the query performance.

Make the cache configurable by the new server settings:
- `text_index_header_cache_policy`
- `text_index_header_cache_size`
- `text_index_header_cache_max_entries`
- `text_index_header_cache_size_ratio`

Additionally, the header cache can be enabled via a new setting
`use_text_index_header_cache`. By default, cache is disabled.

ahmadov committed:
add unit test

ahmadov committed:
add docs for the text inder header cache

Ergus commented on src/Interpreters/Context.cpp:
```diff
@@ -3810,6 +3811,43 @@ void Context::clearTextIndexDictionaryBlockCache() const
+
+    /// Clear the cache without holding context mutex to avoid blocking context for a long time
+    if (cache)
+        cache->clear();
```
You could be calling clear and `getTextIndexHeaderCache` at the same time? Or is there some other protection?

Ergus commented on src/Storages/MergeTree/MergeTreeIndexText.cpp:
```diff
@@ -280,22 +270,62 @@ ColumnPtr deserializeTokensFrontCoding(ReadBuffer & istr, size_t num_tokens)
+                load_header);
+
+        return load_header();
+    };
```
I think that the outer lambda is not needed, just an if-else could do the same much simpler.

ahmadov replied:
in L#323 the `load_header` function is passed to the cache interface but it's only executed when an entry is not in cache. So, like that I tried to avoid duplication with if/else

Ergus commented on src/Storages/MergeTree/MergeTreeIndexText.h:
```diff
@@ -195,6 +195,36 @@ struct DictionaryBlock : public DictionaryBlockBase
+    size_t num_tokens;
+    BloomFilter bloom_filter;
+    DictionarySparseIndex sparse_index;
+};
```
Very minor: Classes like this I prefer to declare as nested classes, so the names get shorter and their purpose is more clear. But that's a personal preference.

Ergus reviewed (changes requested):
In general LGTM. However, I would request the lock holding change or just check if it is needed.

ahmadov commented on src/Core/Defines.h:
```diff
@@ -104,6 +104,10 @@ static constexpr auto DEFAULT_TEXT_INDEX_DICTIONARY_BLOCK_CACHE_POLICY = "SLRU";
+static constexpr auto DEFAULT_TEXT_INDEX_HEADER_CACHE_POLICY = "SLRU";
+static constexpr auto DEFAULT_TEXT_INDEX_HEADER_CACHE_MAX_SIZE = 1_GiB;
+static constexpr auto DEFAULT_TEXT_INDEX_HEADER_CACHE_SIZE_RATIO = 0.5;
+static constexpr auto DEFAULT_TEXT_INDEX_HEADER_CACHE_MAX_ENTRIES = 100'000;
```
I'd say mostly random but also estimated in a way on a smallest system (4GB or 8GB memory) system it takes up to 25%. But they are configurable

Ergus commented on src/Storages/MergeTree/MergeTreeIndexConditionText.h:
```diff
@@ -134,6 +138,10 @@ class MergeTreeIndexConditionText final : public IMergeTreeIndexCondition, WithC
+    /// Using text index header can be enabled to reduce I/O
+    bool use_header_cache;
+    /// Instance of the text index dictionary block cache
+    TextIndexHeaderCache * header_cache;
```
This is a very legitimate use case for shared pointer as storing a pointer that could potentially be deleted (or reassigned the shared content) could be a bit dangerous, but ok.

Ergus commented on src/Storages/MergeTree/TextIndexCache.h:
```diff
@@ -100,7 +105,44 @@ class TextIndexDictionaryBlockCache : public CacheBase<UInt128, TextIndexDiction
+            ProfileEvents::increment(ProfileEvents::TextIndexHeaderCacheHits);
+        return std::move(cache_entry);
+    }
+};
```
We have a bit of both in the code. It is true that forward declaring them is not possible, but that could be a feature not an issue. ;)

ahmadov replied:
f59b5ad50ae9aa50f290c8324131a8c9a6ea81ba

ahmadov committed:
address comments

ahmadov committed:
enable header cache in text index perf tests

Ergus commented on src/Interpreters/Context.cpp:
```diff
@@ -3810,6 +3811,43 @@ void Context::clearTextIndexDictionaryBlockCache() const
+std::shared_ptr<TextIndexHeaderCache> Context::getTextIndexHeaderCache() const
+{
+    SharedLockGuard lock(shared->mutex);
+    return shared->text_index_header_cache;
```
This lock is protecting the variable initialization in the context. No it's internal state. For internal state there is another internal lock, so the `shared->mutex` doesn't need to be hold.

Ergus commented:
> > Maybe if it is possible, it worth to add a performance test so you can prove that it actually makes queries faster.
> 
> ah, that's a good point. I forgot to enable it in the perf tests.

But with the current test as it is written now, is not designed to compare if the new cache accelerates the queries.

IIUC the test spends more time checking the system logs than in queries execution. Also to check how the caches accelerate the queries we need at least 3 executions:
1. With cache disabled (warmup all other caches)
2. With new cache enable first run (warmup new cache)
3. With cache enable hot run

george-larionov commented on src/Storages/MergeTree/MergeTreeIndexText.h:
```diff
@@ -195,6 +186,45 @@ struct DictionaryBlock : public DictionaryBlockBase
+    }
+
+private:
+    size_t num_tokens;
```
Nit: I would personally feel better if this was init-ed to something here

george-larionov reviewed (approved):
This is not crucial, but i would prefer if you could add some more tests for the cache parameters. (Overall, LGTM)

ahmadov committed:
add missing drop view in unit test

also remove `use_text_index_header_cache` parameter from the perf test
because the master branch does not support it yet.
